### PR TITLE
[Parse][IDE] Various parser fixes for code completion

### PR DESCRIPTION
**Pull request by nathawes**
When the parsed expression contained a code completion expression we weren't preserving it in the AST in several cases: ObjC key path expressions, throw statements,  catch clauses, and default arg initializers. This PR removes the early exits, which preserves the completion context in these cases, and doesn't seem to regress anything. Default arg initializers were also not being type-checked in `typeCheckContextAt()`.

The parser was also only adding an implicit return in single expression abstract function decl bodies when parsing via `parseAbstractFunctionBody()`, and not via `parseAbstractFunctionBodyDelayed()`. This PR moves the transformation down into `parseAbstractFunctionBodyImpl()` that is called by both.

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@swift-ci Please Test Source Compatibility

File: lib/Parse/ParseStmt.cpp [hunk 1]
```diff
@@ -899,9 +899,6 @@ ParserResult<Stmt> Parser::parseStmtThrow(SourceLoc tryLoc) {
 
   ParserResult<Expr> Result = parseExpr(diag::expected_expr_throw);
 
-  if (Result.hasCodeCompletion())
-    return makeParserCodeCompletionResult<Stmt>();
-
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp:904]: We need to propagate `hasCodeCompletion()`ness in `if (tryLoc.isValid() && exprLoc.isValid())` branch as well.
 how about something like:
```cpp
  bool hasCodeCompletion = Result.hasCodeCompletion();

  if (Result.isNull())
    Result = ...

  if (tryLoc.isValid() && exprLoc.isValid()) {
    Result = ...
  }

  if (hasCodeCompletion)
    Resullt.setHasCodeCompletion()
```

Comment (rintaro) [reply lib/Parse/ParseStmt.cpp:904]: This affects something like this invalid code:
```swift
try throw something().<HERE>
```

**Comment by rintaro** [comment]
@hamishknight @CodaFi Do you know when `parseAbstractFunctionBodyDelayed()` is actually called in normal compilations? For primary files delayed parsing is disabled. For non-primary files parsing function body is not needed at all I think.

**Comment by hamishknight** [comment]
@rintaro IIRC it may currently be triggered by `SimpleDidSetRequest`, which needs to walk an observer's body to correctly compute its interface type

**Comment by nathawes** [comment]
@swift-ci please test

File: lib/Parse/ParseStmt.cpp (hunk 1 above)
Comment (nathawes) [reply lib/Parse/ParseStmt.cpp:904]: Good catch! Updated it as suggested and added that as a test case. The code change doesn't seem to affect the completions produced in that case, but I figured it's probably still good to have the tests include it for code coverage.

**Review by rintaro — APPROVED** [review]

File: lib/Parse/ParseStmt.cpp (hunk 1 above)
Comment (rintaro) [reply lib/Parse/ParseStmt.cpp:904]: For completions inside function bodies, parser doesn't rely on `hasCodeCompletion()`. So it shouldn't be affected.

I believe 
```
do {
  try throw something().#HERE#
} catch _ {}
```
at top-level shows the effect.

Comment (nathawes) [reply lib/Parse/ParseStmt.cpp:904]: Thanks - let me update the test case.

**Comment by nathawes** [comment]
@swift-ci test and merge

File: lib/Parse/ParseStmt.cpp (hunk 1 above)
Comment (nathawes) [reply lib/Parse/ParseStmt.cpp:904]: done!

**Pull request merged**